Support for Nordic nRF52810 / S112

andrewleech opened the pull request:
### Description

Nordic have a newer smaller addition to the NRF52 family, the NRF52810

We've been told by a rep they're actively trying to push customers away from NRF51 to this new part as it's often cost-competitive to the nRF51's and is supported by current SDK's.

https://www.nordicsemi.com/eng/Products/nRF52810
https://www.nordicsemi.com/eng/Products/S112-SoftDevice

It supports newer features such as the BLE 5 with LE Secure Connections, but does have a reduced peripheral set.

In terms of bringing support into mbed, it has some compoatibility issues due to nordic dropping most of the legacy api interfaces. Eg. there's no UART only UARTE, no SPI just SPIM and SPIS, etc

This does make it quite a bit harder to migrate from nRF51 to this part as opposed to the original nRF52832, but I digress.

This PR is currently only a start towards adding support, there's a fair bit to go I suspect (does not yet compile). 
Much of what I've done so far might be quite naive too, I'm very open to suggestions (and/or assistance) in finishing the support in a clean fashion.

I specifically wanted to get the PR started to get some early feedback regarding integration issues and to check no-one else is also working on this already? 
I definitely don't want to duplicate effort unnecessarily... @marcuschangarm ?

Cheers,
Andrew

### Pull request type

    [ ] Fix
    [ ] Refactor
    [x] New target
    [ ] Feature
    [ ] Breaking change

pi-anl committed:
Basic target support for NRF52810 added

0xc0170 commented:
The code looks fine. As stated, this needs some more work ,might be good to have this on feature branch, @ashok-rao @marcuschangarm to review

andrewleech commented:
I've progressed with this a lot over the week, not cleaned up to push though. It compiles, not yet running.  There's a few limiting issues however. 

There are bugs in sdk14. 
As the nrf52810 no longer has abstractions like nrf_spi you have to use nrf_spis and nrf_spim, which are enabled individually in the SDK config header. If you want to have spim0 and spis0 enabled in the config (for example) and configure at runtime you need to enable the nrf peripheral sharing. That's all fine, except the peripheral sharing file has nrf_spi hardcoded so you have a compile fail on the ntf52810 as soon as you enable peripheral sharing. Nordic don't have any examples on this part with spim or peripheral sharing... as I've found in the past they clearly don't even try to compile their SDK in all their "supported" configurations before a release, let alone test. 

The file with this bug in it no longer exists in sdk15...
As I'm sure you're aware though, nordic (again) did a massive restructure with sdk15 so it's no trivial task to just update, and I presume with their rebuild from the ground up they'll have introduced numerous new bugs, as well as the softdevice for the 810 needing a few for kB of flash for barely any reason. 

Next problem is code size, the 810 doesn't have much left after the softdevice and Nordic SDK are compiled in. 
Currently, mbed_tls is included in a couple of feature ble files, where I don't think I'd needs to be? Basically TLS and another crypto take a good 30k this chip just doesn't have, so I'll need to add some configs to exclude them. No big issue, they're currently just mbedignored. 
I'm not sure yet if there will be room to have the rtos and events.... Currently while compiled with small lib I've got 20k flash free, so will probably be fine. 

I should have a somewhat working initial implementation early next week, without spi or twi support.

ashok-rao commented on targets/targets.json:
```diff
@@ -3592,6 +3592,64 @@
+            }
+        },
+        "OUTPUT_EXT": "hex",
+        "is_disk_virtual": true,
```
@andrewleech : planning to add any post build scripts for combining soft device here?

andrewleech replied:
I thought that's done already by the mbed build tooling, the softdevice is treated as the bootloader.

ashok-rao replied:
@marcuschangarm : has this already landed? can you please confirm?

andrewleech replied:
The printout during compilation certainly appears so, after linking if reports it's merging the s112 into the region before the app, it already gets the correct flash start address and length (both 0x18000) from the definitions and passes them to the linker.

marcuschangarm replied:
Yes, the bootloader changes are in.

ashok-rao commented on targets/targets.json:
```diff
@@ -3592,6 +3592,64 @@
+        "public": false,
+        "detect_code": ["1101"],
+        "program_cycle_s": 6,
+        "bootloader_supported": true
```
Can you add a "device_name" as well? required for DAPLink and other stuff..

andrewleech replied:
Sure, looks simple enough.

ashok-rao commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52810/device/system_nrf52.c.bu:
```diff
@@ -0,0 +1,365 @@
+/*
```
Question to @0xc0170 and @marcuschangarm : Why is this file required?

andrewleech replied:
Sorry, that's not meant to be checked in

andrewleech commented:
I've got a fair bit more to push to this, bit some of it's currently breaking other things so needs cleaning first. 

I've got my local changes all compiling however it doesn't run, with debugger it doesn't even get to the system_nrf52.c initial startup functions. The same codebase compilef for nrf52_dk runs fine so I've got something wrong with the linker, startup assembly or the s112 softdevice.

ashok-rao commented:
Which toolchain are you using? I can suggest to leave the soft device for now, just to boot the device up..you can add in the soft device at a later stage as part of this PR once you have other things in place..you might want to check the *.sct or *.ld based on your toolchain & the device's memory map..

andrewleech commented:
I'm using gcc. Good point about trying without the SD... I've never run an nrf without it but yes it does make sense to simplify the bring up.

marcuschangarm commented:
@andrewleech Thank you for the effort! I just got back in the office. I'll take a look once I've settled in.

cmonr commented:
@andrewleech Would you mind rebasing while @marcuschangarm takes a look at the PR?

cmonr commented:
@marcuschangarm Mind taking a look at this?

marcuschangarm commented:
@cmonr It's on my list! 😄

andrewleech commented:
Yeah no rush, I've been pulled back onto a second project for a bit so it'll probably be a week or more before I continue myself.

cmonr commented:
@marcuschangarm @andrewleech Any update?

pi-anl committed:
Enforce NRF_SDH_BLE_CENTRAL_LINK_COUNT = 0 for S112

cmonr commented:
Let us konw when this is ready to be re-reviewed (Aka remove WIP from the title).

pi-anl committed:
Remove macros that are now set in config system

pi-anl committed:
Add config entry for S112 softdevice.develop_in_nrf52832 for running nRF52810 code on nRF52832 parts.

cmonr commented:
@andrewleech I take it this still a WIP?

andrewleech commented:
I should be getting more time to do further work on this in a couple of weeks, however I believe basic functionality is working at this point. 

There's no spi or twi support at this stage due to nordic removing the older interface to these in favour of spim/spis etc, however I'm not going to be able to work on those drivers myself as our current project didn't use these. 
Thankfully the UART driver has already been migrated to uarte so that'll be up and running, though it clearly needs conflict resolution at the moment.

cmonr commented:
@ashok-rao @marcuschangarm Mind taking another look?

ashok-rao commented:
Changes look fine to me @andrewleech .. I see you've added startup for GCC as well (@marcuschangarm I've not reviewed this, could you please do this?).. Does this now compile with GCC then?

andrewleech commented:
Yes, it compiles and runs with gcc. This is the only compiler I've tested with.

cmonr commented on features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_MCU_NRF51822/source/nRF5xGap.cpp:
```diff
@@ -246,7 +246,7 @@ ble_error_t nRF5xGap::stopScan(void) {
     return BLE_STACK_BUSY;
 }
-#endif
+#endif // !defined(TARGET_MCU_NRF51_16K_S110) && !defined(TARGET_MCU_NRF51_32K_S110) && !defined(S112)
```
:100:

cmonr commented on features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_MCU_NRF51822/source/nRF5xGap.cpp:
```diff
@@ -274,6 +274,8 @@ ble_error_t nRF5xGap::stopAdvertising(void)
 }
 
+/* Central role is not supported by S112, return BLE_ERROR_NOT_IMPLEMENTED */
+#ifndef S112
```
Does this #ifdef still matter if NRF_SDH_BLE_CENTRAL_LINK_COUNT is being set to 0?

andrewleech replied:
I think so, in general the `#ifndef S112`'s were added wherever the stack tried to call functions which nordic simply don't define for the 810

cmonr commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52810/device/system_nrf52810.c:
```diff
@@ -0,0 +1,394 @@
+static bool errata_136(void);
+
+/* Helper functions for Errata workarounds in nRF52832 */
+#if defined (DEVELOP_IN_NRF52832)
```
Are these workarounds still needed for this part?

andrewleech replied:
Nordic don't build a dev kit for the 810, officially they say to use the normal 832 board but compile your 810 code with this `DEVELOP_IN_NRF52832` define set which allows 810 code to run on 832

I've linked this define through to a mbed config to expose it as such

cmonr commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/object_owners.c:
```diff
@@ -21,10 +21,14 @@
 
-static void * nordic_spi2c_owners[SPI2C_INSTANCES] = { NULL, NULL, NULL };
 
+static void * nordic_spi2c_owners[SPI2C_INSTANCES] = { NULL, NULL, NULL };
```
Just a nit, but was this line change intended?

andrewleech replied:
No it wasn't, fixing now

cmonr commented:
Just a few comments. Will start CI once this is rebased, and comments answered, if needed.
But will hold off on merging until someone more familiar with the NRF parts OKs the PR.

cmonr commented:
Going to try something. Something weird happened with GitHub and not detecting that a rebase happened...

andrewleech commented:
Sorry the first push before wasn't actually a rebase, I just fixed the deleted line. I probably should have done it as part of a rebase... but now it's been rebased
And yeah, that one should have been detected as a rebase

After the rebase though it doesn't compile, changes in the serial_api need to be fixed for 810. 
I won't be able to get to this until next week.

cmonr commented:
AH, I thought GitHub was going crazy for a moment there.

Thanks for the heads up. FYI, our code freeze for 5.10 will be at the end of the month, and would love to get this and your other PRs in by then.

andrewleech commented:
Good to know thanks. Hopefully my other one is straightforward enough to not cause issues.

This 810 work is supposed to be highest priority at the start of next sprint (next week) so hopefully it'll get sorted then quickly enough, presuming priorities don't get changed on me again.

pi-anl committed:
nrf52810: Add target to run nRF52810 code on NRF52_DK as recommended by Nordic
http://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v14.1.0%2Fnrf52810_user_guide.html

cmonr commented:
@ashok-rao @pan- @marcuschangarm When y'all get a chance.

cmonr commented:
@andrewleech Mind taking a look at the Travis Docs issue?

cmonr commented:
@ashok-rao @pan- @marcuschangarm When y'all get a chance.

pi-anl committed:
Update nvic settings in linker/startup for IAR and ARD_STD compilers
Fix assembly extension (lower to upper case)

andrewleech commented:
I've been using this branch for the past week and it's working well without rtos, using uarte no worries. 

I haven't tested with rtos (my application doesn't fit it), nor have I tested on any compiler other than gcc. 
I've attempted to ensure linker and startup for the other two look correct as compared to the nrf52832 though.

pi-anl committed:
Update nrf52810 linker script to allow configuration of MBED_RAM_START

TacoGrandeTX commented:
@andrewleech - Do you have a 3rd party board with a nRF52810 that you are using for development?  Nordic still doesn't appear to offer one.  We have an effort underway to add support for the Nordic v15 SDK but have not taken this device into account.

andrewleech commented:
I don't know of any off the shelf boards, no. We're using the module on a couple of custom boards for products.

We had a nordic rep here a while ago who pushed us very strongly to use this chip as they're trying to deprecate the nrf51's entirely. Obviously they've stopped supporting them in the SDK's for some time now and want everyone with lower cost applications using the 810. Their efforts would certainly be easier if they just released a DK with this chip on it though.

The S112 softdevice can be used on the nrf52832 chip with the develop_in_nrf52832 config set, in this case there is a lot more free ram/flash than when using the S132 (for BLE peripheral only support). Similarly the S132 can use used on the nrf52810 however it uses most of the flash/ram on the chip so there's not much room for applications.

cmonr commented:
@pan- @marcuschangarm @ashok-rao Mind taking a look?
This PR has been sitting quietly waiting it's turn :)

0xc0170 commented:
Based on the info above. Should this be `NRF52_DK_52810` kept away from targets.json and kept rather private (you would have own config to set this up) as this is custom board? The MCU_ support is fine to have here. However, would be better if there is a board so can be used and tested.

@bulislaw Please review this target addition

andrewleech commented:
I think there's value in the `NRF52_DK_52810` target, though perhaps it should be renamed to `NRF52_DK_S112`. 

The S112 softdevice is useful to run on the nrf52832 if you want some more free resources for application use and the reduced BLE capabilities don't limit you.

The `NRF52_DK_52810` target is also quite crucial if you're developing an application that's targeting the nrf52810 but only fits there in fully optimised compilation, the `NRF52_DK_52810` target lets you compile without optimization and test on the larger device.

And yes, I wish nordic would just make a dev board with the 810 fitted, I'm sure it's hurting the uptake of this newer chip due to confusion over how to develop for it.

marcuschangarm commented:
> I think there's value in the NRF52_DK_52810 target, though perhaps it should be renamed to NRF52_DK_S112.

You should use the config system to change softdevices and not targets.json, otherwise you would have to create duplicates of all the NRF52832 targets. 

See https://github.com/ARMmbed/mbed-os/tree/master/targets/TARGET_NORDIC/TARGET_NRF5x#changing-softdevice

cmonr commented:
It's unfortunate that because this hasn't seen any activity, and we don't see any internal push to pick up the work, I'm going to have to close this due to inactivity.

@andrewleech I hate to see your work go to waste, but maybe one day soon one of us will be able to pick this up in the near future.
